halo2_proofs/src/plonk/circuit.rs

```diff
@@ -1545,6 +1685,472 @@ impl<F: Field> Gate<F> {
+            .gates
+            .iter()
+            .map(|g| GateV2Backend {
+                name: g.name.clone(),
+                constraint_names: g.constraint_names.clone(),
+                polys: g.polys.clone().into_iter().map(|e| e.into()).collect(),
+            })
+            .collect(),
+        permutation: cs.permutation.clone(),
+        lookups: cs
+            .lookups
+            .iter()
+            .map(|l| lookup::ArgumentV2 {
+                name: l.name.clone(),
+                input_expressions: l
+                    .input_expressions
+                    .clone()
+                    .into_iter()
+                    .map(|e| e.into())
+                    .collect(),
+                table_expressions: l
+                    .table_expressions
+                    .clone()
+                    .into_iter()
+                    .map(|e| e.into())
+                    .collect(),
+            })
+            .collect(),
+        shuffles: cs
+            .shuffles
+            .iter()
+            .map(|s| shuffle::ArgumentV2 {
+                name: s.name.clone(),
+                input_expressions: s
+                    .input_expressions
+                    .clone()
+                    .into_iter()
+                    .map(|e| e.into())
+                    .collect(),
+                shuffle_expressions: s
+                    .shuffle_expressions
+                    .clone()
+                    .into_iter()
+                    .map(|e| e.into())
+                    .collect(),
+            })
+            .collect(),
+        general_column_annotations: cs.general_column_annotations.clone(),
+    };
+    let preprocessing = PreprocessingV2 {
+        permutation: assembly.permutation,
+        fixed,
+    };
+
+    Ok((
+        CompiledCircuitV2 {
+            cs: cs2,
+            preprocessing,
+        },
+        config,
+        cs,
+    ))
+}
+
+impl<F: Field> ConstraintSystemV2Backend<F> {
+    pub(crate) fn collect_queries(
+        &self,
+    ) -> (
+        Queries,
+        Vec<Gate<F>>,
+        Vec<lookup::Argument<F>>,
+        Vec<shuffle::Argument<F>>,
+    ) {
+        let mut queries = QueriesMap {
+            advice_map: HashMap::new(),
+            instance_map: HashMap::new(),
+            fixed_map: HashMap::new(),
+            advice: Vec::new(),
+            instance: Vec::new(),
+            fixed: Vec::new(),
+        };
+
+        let gates: Vec<_> = self
+            .gates
+            .iter()
+            .map(|gate| Gate {
+                name: gate.name.clone(),
+                constraint_names: gate.constraint_names.clone(),
+                polys: gate
+                    .polynomials()
+                    .iter()
+                    .map(|e| queries.as_expression(e))
+                    .collect(),
+                queried_selectors: Vec::new(), // Unused?
+                queried_cells: Vec::new(),     // Unused?
+            })
+            .collect();
```

Comment:
This can probably become an associated fn of `self` such that we split this fn into multiple calls to others which will be easier to test/read.

Reply:
Done in https://github.com/privacy-scaling-explorations/halo2/pull/243/commits/20c16ddf4272f1e92decc514a55d48aafbf0cea0